### ISSUE TITLE
CompressBlock error returns are confusing

### ISSUE THREAD

**Issue by greatroar** [pierrec/lz4#91]
CompressBlock and CompressBlockHC docs currently state:

```
// If the destination buffer size is lower than CompressBlockBound and
// the compressed size is 0 and no error, then the data is incompressible.
//
// An error is returned if the destination buffer is too small.
```

This was introduced to fix #71, but I find it confusing. What is the difference between incompressible and the destination buffer being too small?

**Comment by pierrec** [comment]
In certain cases, the data to be compressed produces a result bigger or equal to the original (uncompressed) input. Although the buffer might be large enough to hold it, we may need to know about it (see the case in LZ4 stream where a block may not be compressed).

**Comment by greatroar** [comment]
I assume you mean https://github.com/pierrec/lz4/blob/4e086a0fe3ed5ba5ec324cbc6b3ecb16691d03f9/internal/lz4stream/frame.go#L277

I see what the code is doing (I was planning to do something similar in a protocol that can send compressed or uncompressed LZ4 blocks), but not why the distinct error conditions are ever needed. The error return value is ignored.

Also, TestIssue71 seems to contradict the docs, or at least my understanding of them. It passes a buffer one byte smaller than CompressBlockBound, but then expects an error. Yet if the buffer size is < CompressBlockBound and the data doesn't fit, shouldn't it be considered incompressible?

**Comment by pierrec** [comment]
If you pass in a buffer which size is >= CompressBlockBound, then you expect the data to be always compressible.
If the size is less, then either it is uncompressible or the buffer is really too small to hold the compressed data.

**Comment by greatroar** [comment]
Ok, I hope I get it now. Assuming len(dst) < CompressBlockBound, if n is the actual size of the compressed data,

* "compressible" means n < len(src),
* a non-nil error means: lz4 can compress this, but you need pass a larger buffer, and
* a nil error means: a larger buffer would not help, except if it were of size CompressBlockBound.

Is that right?

**Comment by pierrec** [comment]
> 
> 
> Ok, I hope I get it now. Assuming len(dst) < CompressBlockBound, if n is the actual size of the compressed data,
> 
>     * "compressible" means n < len(src),

yes

> 
>     * a non-nil error means: lz4 can compress this, but you need pass a larger buffer, and

*may* be able to compress it

> 
>     * a nil error means: a larger buffer would not help, except if it were of size CompressBlockBound.

correct

> 
> 
> Is that right?

**Comment by greatroar** [comment]
Fixed by 0004c74bab2e34a90c2045722116673cd8318497.
